noise/README.md

```diff
@@ -0,0 +1,400 @@
+    keys they encounter and check whether the same party presents the same public
+    key in the future.
+    
+And also:
+
+    * Static key reuse: A static key pair used with Noise should be used with a single
+    hash algorithm. The key pair should not be used outside of Noise, nor with
+    multiple hash algorithms. It is acceptable to use the static key pair with
+    different Noise protocols, provided the same hash algorithm is used in all of
+    them. (Reusing a Noise static key pair outside of Noise would require extremely
+    careful analysis to ensure the uses don't compromise each other, and security
+    proofs are preserved).
+
+All libp2p peers possess a cryptographic keypair which is used to [derive their
+peer id][peer-id-spec], which we will refer to as their "identity keypair." To
+avoid potential static key reuse, and to allow libp2p peers with any type of
+identity keypair to use Noise, noise-libp2p uses a separate static keypair for
+Noise that is distinct from the peer's identity keypair.
+
+To authenticate the static Noise key, peers include a [noise-libp2p handshake
+payload](#the-noise-libp2p-handshake-payload) with their encrypted handshake
+messages. The payload includes the peer's public libp2p identity key, and a
+signature over the static Noise public key, produced with the identity key. This
+simple certificate proves that the possessor of the Noise static keypair is also
+in possession of the private libp2p identity key.
+
+The lifetime of the Noise static keypair is not specified. Implementations MAY
+allow noise-libp2p to be initialized with a stored static key to extend the key
+lifetime beyond a single invocation of a given application, in which case it is
+the application's responsibility to store the key. If no static key is provided
+at initialization, noise-libp2p will generate a new X25519 keypair and use it
+for the lifetime of the process.
+
+### The noise-libp2p Handshake Payload
+
+As mentioned, noise-libp2p uses a static keypair that is distinct from the
+"identity keypair" used to derive a peer's peer id. As a result, we must
+authenticate Noise static keys in a way that proves they belong to the correct
+libp2p peer. To do so, noise-libp2p includes a simple "certificate" in the
+payload of Noise handshake messages.
+
+The noise-libp2p handshake payload is a [protobuf][protobuf] message with the
+following schema:
+
+``` protobuf
+message Identity {
+	bytes pubkey = 1;
+	bytes signature = 2;
+}
+
+```
+
+The `pubkey` field contains the public libp2p identity key, encoded as described
+in the [peer id spec][peer-id-spec].
+
+The `signature` field contains a signature of the Noise static public key,
+produced by the private libp2p identity key. The Noise static key is encoded
+into a byte array according to the rules defined in [section 5 of RFC
+7748][rfc-7748-sec-5] and signed as described in the [peer id
+spec][peer-id-spec].
```

Comment:
We should be using domain separation, so the signed message in another domain (i.e. IPNS) for sure cannot be used in here. See https://github.com/libp2p/specs/blob/master/tls/tls.md#libp2p-public-key-extension